mum-prng.h

```diff
@@ -0,0 +1,157 @@
+   (Intel 4.2GHz i7-4790K), or speed of the generation is about 1120M
+   numbers per sec.  So it is very fast.  */
+
+#ifndef __MUM_PRNG__
+#define __MUM_PRNG__
+
+#include "mum.h"
+
+#ifndef MUM_PRNG_UNROLL
+#define MUM_PRNG_UNROLL 16
+#endif
+
+#if MUM_PRNG_UNROLL < 1 || MUM_PRNG_UNROLL > 16
+#error "wrong MUM_PRNG_UNROLL value"
+#endif
+
+#ifdef __GNUC__
+#define EXPECT(cond, v) __builtin_expect((cond), (v))
+#else
+#define EXPECT(cond, v) (cond)
+#endif
+
+#if defined(__GNUC__) && ((__GNUC__ == 4) &&  (__GNUC_MINOR__ >= 9) || (__GNUC__ > 4))
+#define _MUM_PRNG_FRESH_GCC
+#endif
+
+static struct {
+    int count;
+    void (*update_func) (void);
+    /* MUM PRNG state */
+    uint64_t state[MUM_PRNG_UNROLL];
+} _mum_prng_state;
+
+#if defined(__x86_64__) && defined(_MUM_PRNG_FRESH_GCC)
+/* This code specialized for Haswell generates MULX insns. */
+static inline uint64_t _MUM_TARGET("arch=haswell")
+_mum_avx2(uint64_t v, uint64_t p) {
+    uint64_t hi, lo;
+    __uint128_t r = (__uint128_t)v * (__uint128_t)p;
+    hi = (uint64_t)(r >> 64);
+    lo = (uint64_t)r;
+    return hi + lo;
+}
+
+static void _MUM_TARGET("arch=haswell")
+_mum_prng_update_avx2(void) {
+    int i;
+
+    _mum_prng_state.count = 0;
+    for (i = 0; i < MUM_PRNG_UNROLL - 1; i++)
+        _mum_prng_state.state[i] ^= _mum_avx2(_mum_prng_state.state[i + 1], _mum_primes[i]);
+    _mum_prng_state.state[MUM_PRNG_UNROLL - 1] ^= _mum_avx2(_mum_prng_state.state[0], _mum_primes[MUM_PRNG_UNROLL - 1]);
+}
+#endif
+
+static void _MUM_NOINLINE
+_mum_prng_update(void) {
+    int i;
+
+    _mum_prng_state.count = 0;
+    for (i = 0; i < MUM_PRNG_UNROLL - 1; i++)
+        _mum_prng_state.state[i] ^= _mum(_mum_prng_state.state[i + 1], _mum_primes[i]);
+    _mum_prng_state.state[MUM_PRNG_UNROLL - 1] ^= _mum(_mum_prng_state.state[0], _mum_primes[MUM_PRNG_UNROLL - 1]);
+}
+
+#if defined(__x86_64__) && defined(_MUM_PRNG_FRESH_GCC)
+static inline void
+_mum_prng_setup_avx2(void) {
+    __builtin_cpu_init();
+    if (__builtin_cpu_supports("avx2"))
+        _mum_prng_state.update_func = _mum_prng_update_avx2;
+    else
+        _mum_prng_state.update_func = _mum_prng_update;
+
+}
+#endif
+
+static inline void
+_start_mum_prng(uint32_t seed) {
+    int i;
+
+    _mum_prng_state.count = MUM_PRNG_UNROLL;
+    for (i = 0; i < MUM_PRNG_UNROLL; i++)
+        _mum_prng_state.state[i] = seed + 1;
+#if defined(__x86_64__) && defined(_MUM_PRNG_FRESH_GCC)
+    _mum_prng_setup_avx2();
+#else
+    _mum_prng_state.update_func = _mum_prng_update;
+#endif
+}
+
+static inline void
+init_mum_prng(void) {
+    _start_mum_prng(0);
+}
+
+static inline void
+set_mum_prng_seed(uint32_t seed) {
+    _start_mum_prng(seed);
+}
+
+static inline uint64_t
+get_mum_prn(void) {
+    if (EXPECT(_mum_prng_state.count == MUM_PRNG_UNROLL, 0)) {
+        // NOTE(Oskar): The state is not consistent here so sometimes when calling from as our update_func is null.
```

Comment:
@Paril I made a change here because this library has an internal global state and it's being reset in between compilation units or something, I am not really sure. Monkey patched it for now because I'm unsure what places we need to initialize it to be consistent.